DDD-7.md

```diff
@@ -0,0 +1,421 @@
+The `ThreadPoolExecutor` will be created by factory method [`Executors.newFixedThreadPool​(int nThreads)`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/Executors.html#newFixedThreadPool(int)).
+There will be two thread pools, the one for running tasks in parallel and one for processing CDC event pipelines.
+In context of this document, CDC event pipelines means the chain of record transformations and eventually also record serialization and processing by user-provided `Consumer`.
+Number of the worker thread in the task thread pool will be the same as the number of tasks.
+Number of worker threads for CDC event pipelines will be specified either by the user via configuration or the default value will be used.
+The default value will be the number of the cores of the underlying machine as running the tasks is not CPU intensive.
+
+### Running source tasks in parallel
+
+Kafka Connect `SourceTask` life cycle will be separated into a self-contained task.
+These tasks will be executed in dedicated threads created by [`Executors.newSingleThreadExecutor()`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/Executors.html#newSingleThreadExecutor()).
+In case of `RetriableException` the task will be restarted.
+In other cases, all other tasks will be stopped gracefully and the engine will fail with the task exception.
+
+#### Storing offsets
+
+Running tasks in parallel doesn't seem to be an issue for storing offsets.
+Tasks are created with 1:1 mapping to connector's/DB's partitions (e.g. in case of SQL server each database is a separate partition and for each database will be created one task), so each task should read or write its own key in offset hash map.
+Reading the offsets from multiple tasks therefore shouldn't introduce any concurrency issues and for writing [OffsetStorageWriter](https://github.com/apache/kafka/blob/3.6.0/connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageWriter.java#L65) is explicitly marked as thread-safe.
+
+### Processing CDC events concurrently
+
+Current `DebeziumEngine` API partially delegates event processing to the [`ChangeConsumer`](https://github.com/debezium/debezium/blob/v2.4.0.Final/debezium-api/src/main/java/io/debezium/engine/DebeziumEngine.java#L159) implementation, which is provided by the user.
+More specifically, events are passed to the [`ChangeConsumer.handleBatch()`](https://github.com/debezium/debezium/blob/v2.4.0.Final/debezium-api/src/main/java/io/debezium/engine/DebeziumEngine.java#L167) in a batch manner, as a `List`.
+This makes sense as in many cases events are submitted to another system and this is usually more efficient to do in a batch.
+However, this prevents us from making a complete event processing pipeline, which would be run in a dedicated thread.
```

Comment:
Since you mentioned community feedback, here's one :smile: Apologies if this is outside scope of this design document.

I assume that a large chunk of the time an event uses from source to sink is spent travelling from Debezium to the sink (and the response back to Debezium), especially if the sink is physically located far away from where Debezium is running.

For sinks that maintain ordering of events within partitions, like Kafka and Azure Event Hubs, it seems like one could gain performance if the sink adapter grouped a batch by destination topic, then by partition id, and sent those groups in parallel to the sink? I guess is would depend on batch size and latency between Debezium and sink.

Reply:
hi @aleksandervalle , thanks for the comment. Yes, I totally agree, however, this is not anything we can do in the engine itself as it doesn't know how the sink will look like. This needs to be done in Debezium server, namely in concrete sink implementations.